[JIT] Enable conditional chaining for Intel APX

anthonycanino opened the pull request:
# Design

This PR mostly enables existing conditional chaining logic for X86 with the inclusion of APX `ccmp` instruction. Currently, the optimization must be explicitly enabled via `DOTNET_JitEnableApxConditionalChaining=1`. 

# Testing

Note: The testing plan for APX work has been discussed in https://github.com/dotnet/runtime/pull/106557, please refer to that PR for details, only results and comments will be posted in this PR. Results posted below.

anthonycanino committed:
Enable conditional compare chaining for AMD64.

anthonycanino commented:
# 1. Intel SDE Testing

Test run with SDE:

![base](https://github.com/user-attachments/assets/a5709770-e40d-4e9d-8747-dbb698fb8a5e)

Test run with SDE with `DOTENT_JitEnableApxConditionalChaining=1`

![diff](https://github.com/user-attachments/assets/5b2302ee-6442-4785-bd27-2dc5b37bdb2b)

2. SuperPMI results

Diffs are based on <span style="color:#1460aa">2,635,272</span> contexts (<span style="color:#1460aa">1,050,818</span> MinOpts, <span style="color:#1460aa">1,584,454</span> FullOpts).

<span style="color:#d35400">MISSED</span> contexts: <span style="color:#d35400">2,984 (0.11%)</span>

Base JIT options: JitBypassApxCheck=1

Diff JIT options: JitBypassApxCheck=1;JitEnableApxIfConv=1


<details>
<summary>Overall (<span style="color:green">-169,140</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|PerfScore in Diffs
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|42,216,437|<span style="color:green">-6,257</span>|<span style="color:green">-9.13%</span>|
|benchmarks.run.windows.x64.checked.mch|8,860,704|<span style="color:green">-21,878</span>|<span style="color:green">-4.99%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|35,294,983|<span style="color:green">-25,089</span>|<span style="color:green">-8.82%</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|12,613,813|<span style="color:green">-20,816</span>|<span style="color:green">-4.88%</span>|
|coreclr_tests.run.windows.x64.checked.mch|389,370,227|<span style="color:green">-11,578</span>|<span style="color:green">-8.49%</span>|
|libraries.crossgen2.windows.x64.checked.mch|44,888,851|<span style="color:red">+1,338</span>|<span style="color:green">-8.60%</span>|
|libraries.pmi.windows.x64.checked.mch|60,136,361|<span style="color:green">-7,956</span>|<span style="color:green">-9.53%</span>|
|libraries_tests.run.windows.x64.Release.mch|322,952,768|<span style="color:green">-56,484</span>|<span style="color:green">-10.31%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|147,678,294|<span style="color:green">-20,097</span>|<span style="color:green">-6.96%</span>|
|realworld.run.windows.x64.checked.mch|10,242,976|<span style="color:green">-451</span>|<span style="color:green">-6.47%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|4,496,305|<span style="color:red">+128</span>|<span style="color:green">-6.51%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:green">-169,140</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|PerfScore in Diffs
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|23,294,229|<span style="color:green">-6,257</span>|<span style="color:green">-9.13%</span>|
|benchmarks.run.windows.x64.checked.mch|8,860,282|<span style="color:green">-21,878</span>|<span style="color:green">-4.99%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|20,646,767|<span style="color:green">-25,089</span>|<span style="color:green">-8.82%</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|3,214,524|<span style="color:green">-20,816</span>|<span style="color:green">-4.88%</span>|
|coreclr_tests.run.windows.x64.checked.mch|118,145,145|<span style="color:green">-11,578</span>|<span style="color:green">-8.49%</span>|
|libraries.crossgen2.windows.x64.checked.mch|44,887,136|<span style="color:red">+1,338</span>|<span style="color:green">-8.60%</span>|
|libraries.pmi.windows.x64.checked.mch|60,023,468|<span style="color:green">-7,956</span>|<span style="color:green">-9.53%</span>|
|libraries_tests.run.windows.x64.Release.mch|132,262,168|<span style="color:green">-56,484</span>|<span style="color:green">-10.31%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|137,028,529|<span style="color:green">-20,097</span>|<span style="color:green">-6.96%</span>|
|realworld.run.windows.x64.checked.mch|10,018,142|<span style="color:green">-451</span>|<span style="color:green">-6.47%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|4,495,216|<span style="color:red">+128</span>|<span style="color:green">-6.51%</span>|


</div></details>

jakobbotsch commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -320,9 +320,223 @@ GenTree* Lowering::LowerBinaryArithmetic(GenTreeOp* binOp)
+    // by TryLowerConditionToFlagsNode.
+    //
+    //
+    // On X86, a FP compare is implemented as a fallthrough, which requires two flag checks; hence,
+    // we cannot simply get a single output condition to feed into a ccmp. Might be possible to chain
+    // this, but skipping those cases for now
+    GenCondition cond1;
+    if (op2->OperIsCmpCompare() && varTypeIsIntegralOrI(op2->gtGetOp1()) && IsInvariantInRange(op2, tree) &&
+        ProducesPotentialConsumableFlagsForCCMP(op1) && TryLowerConditionToFlagsNode(tree, op1, &cond1))
```
I think it would be preferable to get rid of `ProducedPotentialConsumableFlagsForCCMP` and add an argument to `TryLowerConditionToFlagsNode` about whether it is allowed to lower to a condition that requires multiple flags checks. Otherwise we end up having to keep `ProducedPotentialConsumableFlagsForCCMP` and `TryLowerConditionToFlagsNode` in sync.

You can use `GenConditionDesc::Get(cond).oper == GT_NONE` to check this condition in the appropriate places in `TryLowerConditionToFlagsNode`.

anthonycanino committed:
Reduce duplication from `optSwitchDetectLikely`.

anthonycanino commented:
Latest superpmi results after rebasing and making some changes. 

Diffs are based on <span style="color:#1460aa">2,669,439</span> contexts (<span style="color:#1460aa">1,084,276</span> MinOpts, <span style="color:#1460aa">1,585,163</span> FullOpts).

<span style="color:#d35400">MISSED</span> contexts: <span style="color:#d35400">1,455 (0.05%)</span>

Base JIT options: JitBypassApxCheck=1

Diff JIT options: JitBypassApxCheck=1;EnableApxConditionalChaining=1


<details>
<summary>Overall (<span style="color:green">-148,632</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|PerfScore in Diffs
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|53,246,741|<span style="color:green">-3,781</span>|<span style="color:green">-3.77%</span>|
|benchmarks.run.windows.x64.checked.mch|8,913,794|<span style="color:green">-21,476</span>|<span style="color:green">-4.44%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|31,189,257|<span style="color:green">-27,274</span>|<span style="color:green">-3.64%</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|12,550,443|<span style="color:green">-20,509</span>|<span style="color:green">-4.42%</span>|
|coreclr_tests.run.windows.x64.checked.mch|409,509,635|<span style="color:green">-8,538</span>|<span style="color:green">-4.23%</span>|
|libraries.crossgen2.windows.x64.checked.mch|39,148,972|<span style="color:red">+4,006</span>|<span style="color:green">-8.70%</span>|
|libraries.pmi.windows.x64.checked.mch|60,485,843|<span style="color:green">-6,789</span>|<span style="color:green">-6.83%</span>|
|libraries_tests.run.windows.x64.Release.mch|328,247,219|<span style="color:green">-52,846</span>|<span style="color:green">-4.07%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|151,291,795|<span style="color:green">-11,562</span>|<span style="color:green">-6.46%</span>|
|realworld.run.windows.x64.checked.mch|11,802,974|<span style="color:green">-131</span>|<span style="color:green">-4.70%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|3,697,229|<span style="color:red">+268</span>|<span style="color:green">-4.41%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:green">-148,632</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|PerfScore in Diffs
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|28,806,967|<span style="color:green">-3,781</span>|<span style="color:green">-3.77%</span>|
|benchmarks.run.windows.x64.checked.mch|8,913,327|<span style="color:green">-21,476</span>|<span style="color:green">-4.44%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|17,455,616|<span style="color:green">-27,274</span>|<span style="color:green">-3.64%</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|3,102,870|<span style="color:green">-20,509</span>|<span style="color:green">-4.42%</span>|
|coreclr_tests.run.windows.x64.checked.mch|125,765,026|<span style="color:green">-8,538</span>|<span style="color:green">-4.23%</span>|
|libraries.crossgen2.windows.x64.checked.mch|39,147,218|<span style="color:red">+4,006</span>|<span style="color:green">-8.70%</span>|
|libraries.pmi.windows.x64.checked.mch|60,372,913|<span style="color:green">-6,789</span>|<span style="color:green">-6.83%</span>|
|libraries_tests.run.windows.x64.Release.mch|132,114,880|<span style="color:green">-52,846</span>|<span style="color:green">-4.07%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|140,638,024|<span style="color:green">-11,562</span>|<span style="color:green">-6.46%</span>|
|realworld.run.windows.x64.checked.mch|11,578,095|<span style="color:green">-131</span>|<span style="color:green">-4.70%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|3,696,221|<span style="color:red">+268</span>|<span style="color:green">-4.41%</span>|


</div></details>

anthonycanino commented:
Note that right now, we have found a small bug that will appear due to some implicitly dependency on `canUseEvexEncoding()` and we are working on a solution. In addition, we needed to patch in some code to enable EGPR for ccmp, which must use extended EVEX to encode EGPR, not REX2 (https://github.com/dotnet/runtime/pull/111072/commits/809f92164123ce7c4470690da2971469f9dfa8c2).

Once we have this ironed out, I can open this as a PR ready for review.

anthonycanino commented:
The follow is the latest superpmi results rebased on main. Note that I am comparing with a patched version of main to force APX (JitBypassApxCheck), so the diff is based on the conditional chaining optimization.

Diffs are based on <span style="color:#1460aa">2,673,047</span> contexts (<span style="color:#1460aa">1,085,228</span> MinOpts, <span style="color:#1460aa">1,587,819</span> FullOpts).

<span style="color:#d35400">MISSED</span> contexts: <span style="color:#d35400">123 (0.00%)</span>

Base JIT options: JitBypassApxCheck=1

Diff JIT options: JitBypassApxCheck=1;EnableApxConditionalChaining=1


<details>
<summary>Overall (<span style="color:green">-140,494</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|PerfScore in Diffs
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|54,183,056|<span style="color:green">-536</span>|<span style="color:red">+2.55%</span>|
|benchmarks.run.windows.x64.checked.mch|8,915,564|<span style="color:green">-21,477</span>|<span style="color:green">-5.22%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|32,755,227|<span style="color:green">-22,078</span>|<span style="color:green">-1.74%</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|12,499,782|<span style="color:green">-20,479</span>|<span style="color:green">-4.91%</span>|
|coreclr_tests.run.windows.x64.checked.mch|409,304,402|<span style="color:green">-10,860</span>|<span style="color:green">-3.56%</span>|
|libraries.crossgen2.windows.x64.checked.mch|38,988,490|<span style="color:red">+3,892</span>|<span style="color:green">-10.34%</span>|
|libraries.pmi.windows.x64.checked.mch|59,843,714|<span style="color:green">-6,897</span>|<span style="color:green">-9.30%</span>|
|libraries_tests.run.windows.x64.Release.mch|329,775,208|<span style="color:green">-50,133</span>|<span style="color:green">-0.12%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|152,213,317|<span style="color:green">-11,713</span>|<span style="color:green">-8.41%</span>|
|realworld.run.windows.x64.checked.mch|11,790,985|<span style="color:green">-138</span>|<span style="color:green">-7.11%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|3,748,084|<span style="color:green">-75</span>|<span style="color:green">-5.71%</span>|


</div></details>

<details>
<summary>FullOpts (<span style="color:green">-140,494</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|PerfScore in Diffs
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|29,887,748|<span style="color:green">-536</span>|<span style="color:red">+2.55%</span>|
|benchmarks.run.windows.x64.checked.mch|8,915,097|<span style="color:green">-21,477</span>|<span style="color:green">-5.22%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|18,993,176|<span style="color:green">-22,078</span>|<span style="color:green">-1.74%</span>|
|benchmarks.run_tiered.windows.x64.checked.mch|3,051,029|<span style="color:green">-20,479</span>|<span style="color:green">-4.91%</span>|
|coreclr_tests.run.windows.x64.checked.mch|125,636,448|<span style="color:green">-10,860</span>|<span style="color:green">-3.56%</span>|
|libraries.crossgen2.windows.x64.checked.mch|38,986,736|<span style="color:red">+3,892</span>|<span style="color:green">-10.34%</span>|
|libraries.pmi.windows.x64.checked.mch|59,730,784|<span style="color:green">-6,897</span>|<span style="color:green">-9.30%</span>|
|libraries_tests.run.windows.x64.Release.mch|133,918,228|<span style="color:green">-50,133</span>|<span style="color:green">-0.12%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|141,524,187|<span style="color:green">-11,713</span>|<span style="color:green">-8.41%</span>|
|realworld.run.windows.x64.checked.mch|11,566,106|<span style="color:green">-138</span>|<span style="color:green">-7.11%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|3,747,076|<span style="color:green">-75</span>|<span style="color:green">-5.71%</span>|


</div></details>

anthonycanino commented:
Attaching the full summary for reference:

[diff_summary.1.md](https://github.com/user-attachments/files/19072905/diff_summary.1.md)

anthonycanino commented:
@jakobbotsch @BruceForstall this is ready for review. It would be helpful if we can discuss any major changes / concerns with respect to this optimization (it will be off by default, and can be tuned at a later point as well).

BruceForstall reviewed (changes requested):
LGTM. Just a few nits: we don't want names in comments, so please fix the "TODO".

jakobbotsch commented on src/coreclr/jit/jitconfigvalues.h:
```diff
@@ -443,6 +443,7 @@ RELEASE_CONFIG_INTEGER(EnableArm64Sve,              "EnableArm64Sve",
 RELEASE_CONFIG_INTEGER(EnableEmbeddedBroadcast,     "EnableEmbeddedBroadcast",   1) // Allows embedded broadcasts to be disabled
 RELEASE_CONFIG_INTEGER(EnableEmbeddedMasking,       "EnableEmbeddedMasking",     1) // Allows embedded masking to be disabled
 RELEASE_CONFIG_INTEGER(EnableApxNDD,                "EnableApxNDD",              0) // Allows APX NDD feature to be disabled
+RELEASE_CONFIG_INTEGER(EnableApxConditionalChaining, "EnableApxConditionalChaining",        0) // Allows APX conditional compare chaining
```
Do we really need a release knob for this?

anthonycanino replied:
I think at the moment, it's good to have as a knob until we are able to tune it on APX hardware (if needed).

anthonycanino commented on src/coreclr/jit/lower.cpp:
```diff
@@ -4422,6 +4425,11 @@ bool Lowering::TryLowerConditionToFlagsNode(GenTree* parent, GenTree* condition,
         GenTree* relopOp2 = relop->gtGetOp2();
 
 #ifdef TARGET_XARCH
+        if (!allowMultipleFlagsChecks && cond->IsFloat())
```
I went through and kept it conservative, but I can try your suggestion. I am away part of this week and all of next week, will give it a chance then.

Are there any other major thoughts or comments?

anthonycanino replied:
I'm looking at the code now and I understand a bit better. Will make the changes and try.

jakobbotsch replied:
> Are there any other major thoughts or comments?

No, I think this starting to look good. Just had a comment about unifying `TryLowerAndOrToCCMP` between the platforms.

jakobbotsch commented on src/coreclr/jit/codegenxarch.cpp:
```diff
@@ -2260,6 +2300,12 @@ void CodeGen::genCodeForTreeNode(GenTree* treeNode)
             // Do nothing; these nodes are simply markers for debug info.
             break;
 
+#if defined(TARGET_AMD64)
+        case GT_CCMP:
+            genCodeForCCMP(treeNode->AsCCMP());
+            break;
+#endif
```
What prevents this from being supported on x86?

BruceForstall replied:
IIUC, APX is only available in Intel 64-bit mode.

anthonycanino committed:
Update src/coreclr/jit/lsrabuild.cpp

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Update src/coreclr/jit/lowerxarch.cpp

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Update src/coreclr/jit/lowerxarch.cpp

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Widen the potential candidates for ccmp folding.

Also lifts GenConditionDesc into CodeGenInterface to better check which
flag lowerings will produce multiple instructions.

jakobbotsch commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -320,9 +320,192 @@ GenTree* Lowering::LowerBinaryArithmetic(GenTreeOp* binOp)
+// Return Value:
+//    false if no changes were made
+//
+bool Lowering::TryLowerAndOrToCCMP(GenTreeOp* tree, GenTree** next)
```
How much does this differ from the arm64 version? Can you just move that version to lower.cpp?

anthonycanino replied:
I've gone through and lifted some common code up.

anthonycanino replied:
I revered the code that lifts `TryLowerAndOrToCCMP` because I can see a number of changes were made.

jakobbotsch replied:
Those changes should be ok for xarch as well. I would still combine them.

anthonycanino replied:
Ok, I've gone ahead and refactored the common code up.

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -4458,6 +4463,16 @@ bool Lowering::TryLowerConditionToFlagsNode(GenTree* parent, GenTree* condition,
             *cond = GenCondition(GenCondition::P);
         }
+
+        if (!allowMultipleFlagsChecks && cond->IsFloat())
+        {
+            const GenConditionDesc& desc = GenConditionDesc::Get(*cond);
+
+            if (desc.oper != GT_NONE)
+            {
+                return false;
+            }
+        }
```
On second thought can we just add the support for this? It seems like it just requires emitting an extra `ccmp` for these cases, similar to the extra cmovs sometimes emitted by `genCodeForSelect`. It will be cleaner than having to move `GenConditionDesc` around and adding architecture-specific limitations.

anthonycanino committed:
Refactor some common code into lower.cpp.

Some code will conflict with latest changes. I've squashed so we
can discuss how to merge in properly.

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -11056,6 +11081,137 @@ bool Lowering::TryLowerAndNegativeOne(GenTreeOp* node, GenTree** nextNode)
+    // For the other operand we can allow more arbitrary operations that set
+    // the condition flags; the final transformation into the flags def is done
+    // by TryLowerConditionToFlagsNode.
+#if defined(TARGET_AMD64)
+    const bool allowMultipleFlagChecks = false;
+#elif defined(TARGET_ARM64)
+    const bool allowMultipleFlagChecks = true;
+#else
+#error Unsupported target
+#endif
```
I think you should pass this as `false` always. In fact there may be a bug around this for arm64 right now.

anthonycanino replied:
Ok, I've made the change.

anthonycanino committed:
Refactored common code out.

anthonycanino commented:
@jakobbotsch I think the failing tests are unrelated. I will grab one more final SPMI results for documentation. Is there any other changes required?

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -4515,6 +4528,18 @@ bool Lowering::TryLowerConditionToFlagsNode(GenTree* parent, GenTree* condition,
 
         *cond = condition->AsCC()->gtCondition;
 
+#ifdef TARGET_XARCH
+        if (!allowMultipleFlagsChecks)
+        {
+            const GenConditionDesc& desc = GenConditionDesc::Get(*cond);
+
+            if (desc.oper != GT_NONE)
+            {
+                return false;
+            }
+        }
+#endif
```
```suggestion
        if (!allowMultipleFlagsChecks)
        {
            const GenConditionDesc& desc = GenConditionDesc::Get(*cond);

            if (desc.oper != GT_NONE)
            {
                return false;
            }
        }
```

jakobbotsch commented on src/coreclr/jit/lowerxarch.cpp:
```diff
@@ -320,9 +320,67 @@ GenTree* Lowering::LowerBinaryArithmetic(GenTreeOp* binOp)
+//    (NOTE: This just has to make the condition be true, i.e., if the condition calls for (SF ^ OF), then
+//    returning one will suffice
+//
+//    TODO-XArch-APX: Revisit this
```
Why this TODO?

anthonycanino replied:
This isn't needed, it is left over from before I understood how the code worked. Internal note.

anthonycanino commented on src/coreclr/jit/switchrecognition.cpp:
```diff
@@ -157,11 +157,14 @@ bool IsConstantTestCondBlock(const BasicBlock* block,
 //    True if the conversion was successful, false otherwise
 //
-bool Compiler::optSwitchDetectAndConvert(BasicBlock* firstBlock)
+bool Compiler::optSwitchDetectAndConvert(BasicBlock* firstBlock, bool testingForConversion)
```
I put this check in when I saw that for x64, with conditonal chaining, an optimized switch statement/jump table was getting converted into a long series of `ccmp`. 

With respect to ARM, I have not dug into that.

jakobbotsch replied:
Can you post the x64 diffs comparing the PR without this and the PR with this?

Ok on arm64, we can try that as a follow up to see how it impacts it.

anthonycanino committed:
Review edits.

anthonycanino commented:
FYI, latest superpmi results with APX and no conditional chaining as the base.

Diffs are based on <span style="color:#1460aa">2,603,000</span> contexts (<span style="color:#1460aa">1,081,295</span> MinOpts, <span style="color:#1460aa">1,521,705</span> FullOpts).

<span style="color:#d35400">MISSED</span> contexts: <span style="color:#d35400">3,511 (0.13%)</span>

Base JIT options: JitBypassApxCheck=1

Diff JIT options: JitBypassApxCheck=1;EnableApxConditionalChaining=1


<details>
<summary>Overall (<span style="color:green">-118,918</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|PerfScore in Diffs
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|57,020,120|<span style="color:green">-1,562</span>|<span style="color:red">+2.56%</span>|
|benchmarks.run.windows.x64.checked.mch|11,381,572|<span style="color:green">-20,107</span>|<span style="color:green">-5.14%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|39,205,286|<span style="color:green">-22,901</span>|<span style="color:green">-3.05%</span>|
|benchmarks.run_pgo_optrepeat.windows.x64.checked.mch|8,821,041|<span style="color:green">-20,229</span>|<span style="color:green">-5.19%</span>|
|coreclr_tests.run.windows.x64.checked.mch|408,476,326|<span style="color:green">-4,499</span>|<span style="color:green">-13.95%</span>|
|libraries.crossgen2.windows.x64.checked.mch|38,986,902|<span style="color:red">+3,959</span>|<span style="color:green">-9.82%</span>|
|libraries.pmi.windows.x64.checked.mch|40,511,830|<span style="color:green">-4,273</span>|<span style="color:green">-9.87%</span>|
|libraries_tests.run.windows.x64.Release.mch|336,563,451|<span style="color:green">-42,717</span>|<span style="color:red">+0.48%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|152,230,624|<span style="color:green">-6,909</span>|<span style="color:green">-8.46%</span>|
|realworld.run.windows.x64.checked.mch|11,556,000|<span style="color:green">-31</span>|<span style="color:green">-6.77%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|4,293,239|<span style="color:red">+351</span>|<span style="color:green">-5.89%</span>|



</div></details>

<details>
<summary>FullOpts (<span style="color:green">-118,918</span> bytes)</summary>
<div style="margin-left:1em">

|Collection|Base size (bytes)|Diff size (bytes)|PerfScore in Diffs
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|28,956,036|<span style="color:green">-1,562</span>|<span style="color:red">+2.56%</span>|
|benchmarks.run.windows.x64.checked.mch|11,381,029|<span style="color:green">-20,107</span>|<span style="color:green">-5.14%</span>|
|benchmarks.run_pgo.windows.x64.checked.mch|20,236,412|<span style="color:green">-22,901</span>|<span style="color:green">-3.05%</span>|
|benchmarks.run_pgo_optrepeat.windows.x64.checked.mch|8,820,574|<span style="color:green">-20,229</span>|<span style="color:green">-5.19%</span>|
|coreclr_tests.run.windows.x64.checked.mch|125,227,025|<span style="color:green">-4,499</span>|<span style="color:green">-13.95%</span>|
|libraries.crossgen2.windows.x64.checked.mch|38,985,148|<span style="color:red">+3,959</span>|<span style="color:green">-9.82%</span>|
|libraries.pmi.windows.x64.checked.mch|40,399,141|<span style="color:green">-4,273</span>|<span style="color:green">-9.87%</span>|
|libraries_tests.run.windows.x64.Release.mch|137,971,643|<span style="color:green">-42,717</span>|<span style="color:red">+0.48%</span>|
|libraries_tests_no_tiered_compilation.run.windows.x64.Release.mch|140,977,044|<span style="color:green">-6,909</span>|<span style="color:green">-8.46%</span>|
|realworld.run.windows.x64.checked.mch|11,331,121|<span style="color:green">-31</span>|<span style="color:green">-6.77%</span>|
|smoke_tests.nativeaot.windows.x64.checked.mch|4,292,221|<span style="color:red">+351</span>|<span style="color:green">-5.89%</span>|


</div></details>

anthonycanino committed:
Fix build errors.

BruceForstall commented:
@anthonycanino Please fix the formatting issues.
@jakobbotsch Do you have any more review feedback?

jakobbotsch commented on src/coreclr/jit/lower.cpp:
```diff
@@ -4460,6 +4463,18 @@ bool Lowering::TryLowerConditionToFlagsNode(GenTree* parent, GenTree* condition,
 #endif
 
+#if !defined(TARGET_LOONGARCH64) && !defined(TARGET_RISCV64)
+        if (!allowMultipleFlagsChecks && cond->IsFloat())
```
```suggestion
        if (!allowMultipleFlagsChecks)
```

Nit (the check below also doesn't have this)

jakobbotsch reviewed (approved):
LGTM with a nit.

anthonycanino committed:
Formatting.

BruceForstall commented:
/ba-g azure linux timeouts

anthonycanino commented:
Thanks for all the review efforts @jakobbotsch and @BruceForstall !
